File: Sem2/AI/Lista 4/BoardGames/Reversi/minimax.cpp
#include <bits/stdc++.h>
using namespace std;

/*
black player starts
starting sheme:
...WB...
...BW...
*/

class Reversi{
    const int DIRS[8][2] = {{-1, -1}, {-1, 0}, {-1, 1}, {0, 1},
                            {1, 1},   {1, 0},  {1, -1}, {0, -1}};

    const vector<vector<int>> CELL_WEIGHTS = {{20, -3, 11, 8, 8, 11, -3, 20},
                                              {-3, -7, -4, 1, 1, -4, -7, -3},
                                              {11, -4, 2, 2, 2, 2, -4, 11},
                                              {8, 1, 2, -3, -3, 2, 1, 8},
                                              {8, 1, 2, -3, -3, 2, 1, 8},
                                              {11, -4, 2, 2, 2, 2, -4, 11},
                                              {-3, -7, -4, 1, 1, -4, -7, -3},
                                              {20, -3, 11, 8, 8, 11, -3, 20}};

    const int CORNERS[4][2] = {{0, 0}, {0, 7}, {7, 0}, {7, 7}};

public:
    bool player[8][8];          // active player player's pieces
    bool opponent[8][8];        // opponent's pieces
    bool main_player;           // true if player is main player (else opponent is main player)

    Reversi(){}
    Reversi(bool black_starts){
        reset(black_starts);
    }

    void reset(bool black_starts){
        main_player = true;
        memset(player, 0, sizeof(player));
        memset(opponent, 0, sizeof(opponent));
        if(black_starts){
            opponent[3][3] = opponent[4][4] = true;
            player[3][4] = player[4][3] = true;   
        } else{
            player[3][3] = player[4][4] = true;
            opponent[3][4] = opponent[4][3] = true;   
        }
    }

    /* #region //* setters getters */
    inline void set_player_cell(int x, int y, bool val){
        if(safe(x, y))
        player[x][y] = val;
    }
    inline bool get_player_cell(int x, int y) const {
        return player[x][y];
    }
    inline void set_opponent_cell(int x, int y, bool val){
        if(safe(x, y))
        opponent[x][y] = val;
    }
    inline bool get_opponent_cell(int x, int y) const {
        return opponent[x][y];
    }
    /* #endregion */

    void swap_players(){
        swap(player, opponent);
        main_player = !main_player;
    }

    void make_move(int x, int y){
        set_player_cell(x, y, true);
        for(auto [dx, dy] : DIRS){
            if(can_beat(x, y, {dx, dy})){
                int tx = x + dx, ty = y + dy;
                while(get_opponent_cell(tx, ty)){
                    set_opponent_cell(tx, ty, false);
                    set_player_cell(tx, ty, true);
                    tx += dx;
                    ty += dy;
                }
            }
        }
    }

    inline bool safe(int x, int y) const {
        return x >= 0 && x < 8 && y >= 0 && y < 8;
    } 

    bool can_beat(int x, int y, pair<int, int> dir) const {
        int dx = dir.first, dy = dir.second;
        x += dx, y += dy;

        if(!safe(x, y) || !get_opponent_cell(x, y)) return false;
        while(safe(x, y) && get_opponent_cell(x, y)){
            x += dx;
            y += dy;
        }

        if(safe(x, y) && get_player_cell(x, y)) return true;
        return false;
    }

    vector<pair<int, int>> free_cells() const {
        // long long occupied_cells = (player | opponent);
        vector<pair<int, int>> res;

        for(int x = 0; x < 8; x++){
            for(int y = 0; y < 8; y++){
                if(get_player_cell(x, y) || get_opponent_cell(x, y)) continue;
                for(auto [dx, dy] : DIRS){
                    if(can_beat(x, y, {dx, dy})){
                        res.push_back({x, y});
                        break;
                    }
                }
            }
        }

        return res;
    }

    int result() const {
        int res = 0;
        for(int i = 0; i < 8; i++) for(int j = 0; j < 8; j++) {
            if(get_player_cell(i, j)) res++;
            else if(get_opponent_cell(i, j)) res--;
        }

        return res * (main_player ? 1 : -1);
    }

    inline int calc_ratio(int player, int opponent) const {
        return 100 * ((double)(player - opponent)) / (player + opponent);
    }

    int heuristic_result() const {
        int weighted_sum = 0;
        int player_cells = 0;
        int opponent_cells = 0;
        for(int i = 0; i < 8; i++) for(int j = 0; j < 8; j++) {
            if(get_player_cell(i, j)){
                weighted_sum += CELL_WEIGHTS[i][j];
                player_cells++;
            }
            else if(get_opponent_cell(i, j)){
                weighted_sum -= CELL_WEIGHTS[i][j];
                opponent_cells++;
            }
        }

        int ratio = calc_ratio(player_cells, opponent_cells);
        // if(free_move != 0) weighted_sum *= (1/free_move);


        int player_corners = 0;
        int opponent_corners = 0;
        for(auto [x, y] : CORNERS){
            if(get_player_cell(x, y)) player_corners++;
            else if(get_opponent_cell(x, y)) opponent_corners++;
        }

        int corner_ratio = 0;
        if(player_corners + opponent_corners)
            corner_ratio = calc_ratio(player_corners, opponent_corners);


        int player_close_corners = 0;
        int opponent_close_corners = 0;
        for(auto [x, y] : CORNERS){
            if(not (get_player_cell(x, y) || get_opponent_cell(x, y))){
                for(auto [dx, dy] : DIRS){      // could be more efficient, but quite uglier
                    int tx = x + dx, ty = y + dy;
                    if(not safe(tx, ty)) continue;

                    if(get_player_cell(tx, ty)) player_close_corners++;
                    else if(get_opponent_cell(tx, ty)) opponent_close_corners++;
                }
            }
        }

        int close_corner_value = -12.5 * (player_close_corners - opponent_close_corners);
        
        return ((10*ratio) + (801.724*corner_ratio) + (382.026*close_corner_value) + (78.922*weighted_sum)) * (main_player ? 1 : -1);
    }

    bool terminal() const {
        return free_cells().empty();
    }
    inline bool terminal(vector<pair<int, int>> free_cells) const {
        return free_cells.empty();
    }

    void play(int x, int y){
        make_move(x, y);
        swap_players();
    }

    Reversi gen_next_state(int x, int y){
        Reversi res;
        for(int i = 0; i < 8; i++) for(int j = 0; j < 8; j++) res.player[i][j] = player[i][j];
        for(int i = 0; i < 8; i++) for(int j = 0; j < 8; j++) res.opponent[i][j] = opponent[i][j];
        if(x != -1) res.make_move(x, y);
        res.swap_players();
        return res;
    }

    friend ostream &operator<<(ostream &out, const Reversi &board) {
        for(int i = 0; i < 8; i++){
            for(int j = 0; j < 8; j++){
                out << (board.get_player_cell(i, j) ? 'P' : board.get_opponent_cell(i, j) ? 'O' : '.');
            }
            cout << '\n';
        }
        return out;
    }
};

class AI{
    const int max_depth = 4;
    const int MAX = 1, MIN = 0;
public:
    pair<int, int> get_best_move(Reversi state){
        return decision(state);
    }

    // minimax
    pair<int, int> decision(Reversi state){        // returns best move
        int best_score = INT_MIN;
        pair<int, int> best_move = {-1, -1};
        for(auto [x, y] : state.free_cells()){
            int score = minimax(state.gen_next_state(x, y), max_depth, MIN);
            if(score > best_score){
                best_score = score;
                best_move = {x, y};
            }
        }

        return best_move;
    }

    int minimax(Reversi state, int depth, int player){
        vector<pair<int, int>> free_cells = state.free_cells();

        if(state.terminal(free_cells)){
            Reversi next = state.gen_next_state(-1, -1);
            if(next.terminal()) return state.result();
            else return minimax(next, depth - 1, 1 - player);
        }

        if(depth == 0) return state.heuristic_result();


        int min_score = INT_MAX, max_score = INT_MIN;
        for(auto [x, y] : free_cells){
            int score = minimax(state.gen_next_state(x, y), depth - 1, 1 - player);
            min_score = min(min_score, score);
            max_score = max(max_score, score);
        }

        if(player == MIN){
            return min_score;
        } else{
            return max_score;
        }
    }

    // alpha-beta
    //TODO
};


void say(string what){
    printf("%s\n", what.c_str());
    fflush(stdout);
}

void say(string what, int x, int y){
    printf("%s %d %d\n", what.c_str(), x, y);
    fflush(stdout);
}

int main(){
    srand (time(NULL));
    Reversi game(false);
    AI ai;
    string cmd = "";
    say("RDY");
    
    while(cmd != "BYE"){
        cin >> cmd;

        if(cmd == "UGO"){
            double time_for_move, time_for_game; 
            cin >> time_for_move >> time_for_game;
            game.reset(true);

            auto p = ai.get_best_move(game);
            say("IDO", p.first, p.second);
            game.make_move(p.first, p.second);
            // cerr << game << '\n';
            game.swap_players();
        } else if(cmd == "HEDID"){
            double time_for_move, time_for_game; 
            cin >> time_for_move >> time_for_game;

            int x, y; cin >> x >> y;
            if(x != -1) game.make_move(x, y);
            game.swap_players();
            
            auto p = ai.get_best_move(game);
            say("IDO", p.first, p.second);
            game.make_move(p.first, p.second);
            // cerr << game << '\n';
            game.swap_players();
        } else if(cmd == "ONEMORE"){
            game.reset(false);
            say("RDY");
        }
    }
}

File: Sem2/AI/Lista 4/BoardGames/Reversi/minimax.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-


import copy
import random
import sys

INT_MAX = sys.maxsize
INT_MIN = -sys.maxsize - 1


class Reversi:
    M = 8
    DIRS = [(0, 1), (1, 0),   (-1, 0), (0, -1),
            (1, 1), (-1, -1), (1, -1), (-1, 1)]

    def __init__(self):
        self.board = self.initial_board()
        self.fields = set()
        self.move_list = []
        for i in range(self.M):
            for j in range(self.M):
                if self.board[i][j] is None:
                    self.fields.add((j, i))

    def __hash__(self):
        return hash(tuple([tuple(x) for x in self.board]))

    def initial_board(self):
        B = [[None] * self.M for _ in range(self.M)]
        B[3][3] = 1
        B[4][4] = 1
        B[3][4] = 0
        B[4][3] = 0
        return B

    def draw(self):
        for i in range(self.M):
            res = []
            for j in range(self.M):
                b = self.board[i][j]
                if b is None:
                    res.append('.')
                elif b == 1:
                    res.append('#')
                else:
                    res.append('o')
            print(''.join(res))
        print('')

    def moves(self, player):
        res = []
        for (x, y) in self.fields:
            if any(self.can_beat(x, y, direction, player)
                   for direction in self.DIRS):
                res.append((x, y))
        return res

    def can_beat(self, x, y, dir, player):
        dx, dy = dir
        x += dx
        y += dy
        cnt = 0
        while self.get(x, y) == 1 - player:
            x += dx
            y += dy
            cnt += 1
        return cnt > 0 and self.get(x, y) == player

    def get(self, x, y):
        if 0 <= x < self.M and 0 <= y < self.M:
            return self.board[y][x]
        return None

    def do_move(self, move, player):
        assert player == len(self.move_list) % 2
        self.move_list.append(move)

        if move is None:
            return
        x, y = move
        x0, y0 = move
        self.board[y][x] = player
        self.fields -= set([move])
        for dx, dy in self.DIRS:
            x, y = x0, y0
            to_beat = []
            x += dx
            y += dy
            while self.get(x, y) == 1 - player:
                to_beat.append((x, y))
                x += dx
                y += dy
            if self.get(x, y) == player:
                for (nx, ny) in to_beat:
                    self.board[ny][nx] = player

    def result(self):
        res = 0
        for y in range(self.M):
            for x in range(self.M):
                b = self.board[y][x]
                if b == 0:
                    res -= 1
                elif b == 1:
                    res += 1
        return res

    def terminal(self):
        if not self.fields:
            return True
        if len(self.move_list) < 2:
            return False
        return self.move_list[-1] is None and self.move_list[-2] is None

    def next_state(self, move, player):
        res = copy.deepcopy(self)
        res.do_move(move, player)
        return res


class Minimax(object):
    def __init__(self):
        self.hashmap = dict()

    def cutoff_test(self, state, depth):
        return depth > 2

    def rollout(self, state, player):
        while not state.terminal():
            moves = state.moves(player)
            if len(moves) == 0:
                state = state.next_state(None, player)
            else:
                state = state.next_state(random.choice(moves), player)
            player = 1 - player
        return state.result()

    def heuristic(self, state, player, samples=1):
        res = 0
        for _ in range(samples):
            res += self.rollout(state, player)
        return res / samples

    def min_move(self, state, depth, player, alpha):
        if state.terminal():
            return state.result()
        if self.cutoff_test(state, depth):
            return self.heuristic(state, player)
        
        moves = state.moves(player)
        if len(moves) == 0:
            next_move = state.next_state(None, player)
            return self.max_move(next_move, depth+1, 1-player, INT_MAX)

        min_score = INT_MAX
        for move in moves:
            next_move = state.next_state(move, player)

            score = self.max_move(next_move, depth+1, 1-player, min_score)
            if score < alpha:
                min_score = alpha
                break
            min_score = min(min_score, score)
                
        return min_score

    def max_move(self, state, depth, player, beta):
        if state.terminal():
            return state.result()
        if self.cutoff_test(state, depth):
            return self.heuristic(state, player)
        
        moves = state.moves(player)
        if len(moves) == 0:
            next_move = state.next_state(None, player)
            return self.min_move(next_move, depth+1, 1-player, INT_MIN)
        
        max_score = INT_MIN
        for move in moves:
            next_move = state.next_state(move, player)

            score = self.min_move(next_move, depth+1, 1-player, max_score)
            if score > beta:
                max_score = beta
                break
            max_score = max(max_score, score)

        return max_score

    def minimax(self, state, player):
        moves = state.moves(player)

        if len(moves) == 0:
            return None
        
        max_score = INT_MIN
        best_move = None
        for move in moves:
            next_move = state.next_state(move, player)

            score = self.min_move(next_move, 0, 1-player, max_score)
            if score > max_score:
                max_score = score
                best_move = move

        return best_move


class Player(object):
    def __init__(self):
        self.reset()

    def reset(self):
        self.game = Reversi()
        self.my_player = 1
        self.say('RDY')

    def say(self, what):
        sys.stdout.write(what)
        sys.stdout.write('\n')
        sys.stdout.flush()

    def hear(self):
        line = sys.stdin.readline().split()
        return line[0], line[1:]

    def loop(self):
        CORNERS = { (0,0), (0,7), (7,0), (7,7) }
        minimax = Minimax()

        while True:
            cmd, args = self.hear()
            if cmd == 'HEDID':
                unused_move_timeout, unused_game_timeout = args[:2]
                move = tuple((int(m) for m in args[2:]))
                if move == (-1, -1):
                    move = None
                self.game.do_move(move, 1 - self.my_player)
            elif cmd == 'ONEMORE':
                self.reset()
                continue
            elif cmd == 'BYE':
                break
            else:
                assert cmd == 'UGO'
                assert not self.game.move_list
                self.my_player = 0

            moves = self.game.moves(self.my_player)
            better_moves = list(set(moves) & CORNERS)

            if better_moves:
                move = random.choice(better_moves)
                self.game.do_move(move, self.my_player)
            elif moves:
                move = minimax.minimax(self.game, self.my_player)
                self.game.do_move(move, self.my_player)
            else:
                self.game.do_move(None, self.my_player)
                move = (-1, -1)
            self.say('IDO %d %d' % move)


if __name__ == '__main__':
    player = Player()
    player.loop()
